EventListener/LogoutEventListener.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of the GesdinetJWTRefreshTokenBundle package.
+ *
+ * (c) Gesdinet <http://www.gesdinet.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Gesdinet\JWTRefreshTokenBundle\EventListener;
+
+use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenManagerInterface;
+use Gesdinet\JWTRefreshTokenBundle\Request\Extractor\ExtractorInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\Security\Http\Event\LogoutEvent;
+
+class LogoutEventListener
+{
+    /**
+     * @var RefreshTokenManagerInterface
+     */
+    private $refreshTokenManager;
+
+    /**
+     * @var ExtractorInterface
+     */
+    private $refreshTokenExtractor;
+
+    /**
+     * @var string
+     */
+    private $tokenParameterName;
+
+    /**
+     * @var array
+     */
+    private $cookieSettings;
+
+    /**
+     * @var string
+     */
+    private $logout_firewall_context;
+
+    public function __construct(
+        RefreshTokenManagerInterface $refreshTokenManager,
+        ExtractorInterface $refreshTokenExtractor,
+        string $tokenParameterName,
+        array $cookieSettings,
+        string $logout_firewall_context,
+    ) {
+        $this->refreshTokenManager = $refreshTokenManager;
+        $this->refreshTokenExtractor = $refreshTokenExtractor;
+        $this->tokenParameterName = $tokenParameterName;
+        $this->cookieSettings = array_merge([
+            'enabled' => false,
+            'same_site' => 'lax',
+            'path' => '/',
+            'domain' => null,
+            'http_only' => true,
+            'secure' => true,
+            'remove_token_from_body' => true,
+        ], $cookieSettings);
+        $this->logout_firewall_context = $logout_firewall_context;
+    }
+
+    public function onLogout(LogoutEvent $event): void
+    {
+        $request = $event->getRequest();
+        $current_firewall_context = $request->attributes->get('_firewall_context');
+
+        if ($current_firewall_context !== $this->logout_firewall_context) {
+            return;
+        }
+
+        $tokenString = $this->refreshTokenExtractor->getRefreshToken($request, $this->tokenParameterName);
+        if (null === $tokenString) {
+            $event->setResponse(
+                new JsonResponse(
+                    [
+                        'code' => 422,
+                        'message' => 'No refresh_token found.',
+                    ],
+                    JsonResponse::HTTP_UNPROCESSABLE_ENTITY
```

Comment:
Do we really need to have logout trigger a 422 response if the refresh token isn't part of the request?

Reply:
I'm on the fence. If we don't, the cookie might be cleared on the client-side but the refresh token will not be invalidated on the API side.

Reply:
So early return in the case where there is no refresh token in the request instead of sending a 422 which also has the side effect of invalidating the client cookie, and then only invalidate the cookie if this listener is going to set a 200 response.  Most 4xx responses indicate the client can change something about the request and retry it, but if you invalidate the cookie even when there's no token in the request, then there isn't a way to retry the request.  I think a 400 response in this case might also be more appropriate than 422.